File: Sources/PUPSISPortalApp/Core/ScheduleParser.swift
import Foundation

/// Turns a scraped SIS schedule row into individual class blocks.
///
/// The schedule cell looks like:
///     "1N - BSCS 1-1N - T/F 02:00PM-04:00PM/01:30PM-04:30PM"
/// Days and time ranges are `/`-separated and paired positionally, so the
/// same day twice (`SUN/SUN`) means two blocks that day (Lec then Lab).
enum ScheduleParser {
    /// Parses every scraped row and flattens the result, giving each repeat of
    /// an already-seen subject/day/time an `occurrenceIndex` so duplicate rows
    /// (a literal repeated `<tr>` — a known SIS/DataTables scrape quirk) don't
    /// collide on `ClassSession.id`, which Preferences and SwiftUI identity key
    /// on. A non-duplicate session keeps `occurrenceIndex == 0`, the same `id`
    /// as before this existed.
    static func parse(_ rows: [[String: String]]) -> [ClassSession] {
        var seen: [String: Int] = [:]
        return rows.flatMap(parse).map { session in
            let key = "\(session.subjectCode)-\(session.day.rawValue)-\(session.start)-\(session.end)"
            let index = seen[key, default: 0]
            seen[key] = index + 1
            var copy = session
            copy.occurrenceIndex = index
            return copy
        }
    }

    static func parse(_ row: [String: String]) -> [ClassSession] {
        let line = row["scheduleLine"] ?? ""
        guard let (dayField, timeField) = splitDaysAndTimes(line) else { return [] }

        // Each slash-separated segment is a group; day and time groups pair
        // positionally. A group whose day run expands to more than one day
        // (`TTH`, `MW`) repeats that group's own single time range across
        // every day in it, rather than losing the extra days to a flattened
        // day/range count mismatch.
        let dayGroups = dayField.split(separator: "/").map { tokenizeDays(String($0)) }
        let rangeGroups = timeField.split(separator: "/").compactMap { parseRange(String($0)) }
        guard !dayGroups.isEmpty, !rangeGroups.isEmpty else { return [] }

        let subjectCode = row["subjectCode"] ?? ""
        let description = row["description"] ?? ""
        let faculty = row["faculty"] ?? ""

        let groupCount = min(dayGroups.count, rangeGroups.count)
        return (0..<groupCount).flatMap { index -> [ClassSession] in
            let range = rangeGroups[index]
            return dayGroups[index].map { day in
                ClassSession(
                    subjectCode: subjectCode,
                    description: description,
                    faculty: faculty,
                    day: day,
                    start: range.0,
                    end: range.1
                )
            }
        }
    }

    /// Pulls the "<DAYS> <TIMES>" block out of the schedule line, ignoring
    /// the section prefix (which contains digits and hyphens).
    ///
    /// Deliberately not end-anchored: some rows leak trailing text after the
    /// times (a faculty name that didn't split on `<br>` the way
    /// `SISScraper` expects for most rows) — the section prefix is already
    /// found mid-string, so there's no reason the times must be the last
    /// thing either.
    private static func splitDaysAndTimes(_ line: String) -> (String, String)? {
        let pattern = #"([A-Z]+(?:/[A-Z]+)*)\s+((?:\d{1,2}:\d{2}[AP]M-\d{1,2}:\d{2}[AP]M)(?:/\d{1,2}:\d{2}[AP]M-\d{1,2}:\d{2}[AP]M)*)"#
        guard let regex = try? NSRegularExpression(pattern: pattern),
              let match = regex.firstMatch(in: line, range: NSRange(line.startIndex..., in: line)),
              let dayRange = Range(match.range(at: 1), in: line),
              let timeRange = Range(match.range(at: 2), in: line)
        else { return nil }
        return (String(line[dayRange]), String(line[timeRange]))
    }

    /// Handles both slash-separated codes and runs like "TTH" / "MW".
    private static func tokenizeDays(_ token: String) -> [Weekday] {
        var remainder = Substring(token)
        var days: [Weekday] = []
        outer: while !remainder.isEmpty {
            for (code, day) in Weekday.codes where remainder.hasPrefix(code) {
                days.append(day)
                remainder = remainder.dropFirst(code.count)
                continue outer
            }
            return days.isEmpty ? [] : days // unrecognized tail — keep what parsed
        }
        return days
    }

    /// "02:00PM-04:00PM" -> (840, 960) in minutes from midnight.
    private static func parseRange(_ text: String) -> (Int, Int)? {
        let parts = text.split(separator: "-")
        guard parts.count == 2,
              let start = parseTime(String(parts[0])),
              let end = parseTime(String(parts[1]))
        else { return nil }
        return (start, end)
    }

    private static func parseTime(_ text: String) -> Int? {
        let trimmed = text.trimmingCharacters(in: .whitespaces).uppercased()
        guard trimmed.count >= 6 else { return nil }
        let period = String(trimmed.suffix(2))
        let clock = trimmed.dropLast(2).split(separator: ":")
        guard period == "AM" || period == "PM",
              clock.count == 2,
              var hour = Int(clock[0]),
              let minute = Int(clock[1]),
              (1...12).contains(hour),
              (0..<60).contains(minute)
        else { return nil }
        if period == "PM" && hour != 12 { hour += 12 }
        if period == "AM" && hour == 12 { hour = 0 }
        return hour * 60 + minute
    }
}

File: windows/PUPSISPortal.Core/ScheduleParser.cs
using System.Text.RegularExpressions;

namespace PUPSISPortal.Core;

/// <summary>
/// Turns a scraped SIS schedule row into individual class blocks.
///
/// The schedule cell looks like:
///     "1N - BSCS 1-1N - T/F 02:00PM-04:00PM/01:30PM-04:30PM"
/// Days and time ranges are `/`-separated and paired positionally, so the
/// same day twice (`SUN/SUN`) means two blocks that day (Lec then Lab).
/// </summary>
public static class ScheduleParser
{
    public static List<ClassSession> Parse(Dictionary<string, string> row)
    {
        var line = row.TryGetValue("scheduleLine", out var value) ? value : "";
        var result = SplitDaysAndTimes(line);
        if (result == null)
            return new List<ClassSession>();
        var (dayField, timeField) = result.Value;

        var days = dayField
            .Split('/')
            .SelectMany(token => WeekdayExtensions.TokenizeDays(token))
            .ToList();

        var ranges = timeField
            .Split('/')
            .Select(token => ParseRange(token))
            .Where(r => r.HasValue)
            .Select(r => r!.Value)
            .ToList();

        if (days.Count == 0 || ranges.Count == 0)
            return new List<ClassSession>();

        var subjectCode = row.TryGetValue("subjectCode", out var sc) ? sc : "";
        var description = row.TryGetValue("description", out var d) ? d : "";
        var faculty = row.TryGetValue("faculty", out var f) ? f : "";

        // One day with several ranges means the same day meets more than once;
        // otherwise days and ranges line up one-to-one.
        var count = days.Count == 1 ? ranges.Count : Math.Min(days.Count, ranges.Count);
        var sessions = new List<ClassSession>();

        for (int index = 0; index < count; index++)
        {
            var day = days.Count == 1 ? days[0] : days[index];
            var range = ranges[Math.Min(index, ranges.Count - 1)];
            sessions.Add(new ClassSession
            {
                SubjectCode = subjectCode,
                Description = description,
                Faculty = faculty,
                Day = day,
                Start = range.start,
                End = range.end
            });
        }

        return sessions;
    }

    /// <summary>
    /// Pulls the trailing "<DAYS> <TIMES>" off the end of the schedule line,
    /// ignoring the section prefix (which contains digits and hyphens).
    /// </summary>
    private static (string dayField, string timeField)? SplitDaysAndTimes(string line)
    {
        if (string.IsNullOrEmpty(line))
            return null;

        // Pattern: day codes followed by time ranges
        // Day codes: one or more uppercase letters and slashes, at least two letters
        // Time ranges: one or more "H:MM(AM|PM)-H:MM(AM|PM)" separated by slashes
        const string pattern = @"([A-Z]+(?:/[A-Z]+)*)\s+((?:\d{1,2}:\d{2}[AP]M-\d{1,2}:\d{2}[AP]M)(?:/\d{1,2}:\d{2}[AP]M-\d{1,2}:\d{2}[AP]M)*)\s*$";
        var match = Regex.Match(line, pattern);
        if (!match.Success || match.Groups.Count < 3)
            return null;

        var dayField = match.Groups[1].Value;
        var timeField = match.Groups[2].Value;
        return (dayField, timeField);
    }

    /// <summary>
    /// "02:00PM-04:00PM" -> (840, 960) in minutes from midnight.
    /// </summary>
    private static (int start, int end)? ParseRange(string text)
    {
        var trimmed = text.Trim();
        var parts = trimmed.Split('-');
        if (parts.Length != 2)
            return null;

        var start = ParseTime(parts[0]);
        var end = ParseTime(parts[1]);
        if (start == null || end == null)
            return null;

        return (start.Value, end.Value);
    }

    /// <summary>
    /// Parse a time like "02:00PM" or "12:30AM" to minutes from midnight.
    /// Returns null for unparseable input.
    /// </summary>
    private static int? ParseTime(string text)
    {
        var trimmed = text.Trim().ToUpperInvariant();
        if (trimmed.Length < 6)
            return null;

        var period = trimmed[^2..];
        if (period != "AM" && period != "PM")
            return null;

        var clock = trimmed[..^2].Split(':');
        if (clock.Length != 2)
            return null;

        if (!int.TryParse(clock[0], out var hour) || !int.TryParse(clock[1], out var minute))
            return null;

        if (hour < 1 || hour > 12 || minute < 0 || minute >= 60)
            return null;

        if (period == "PM" && hour != 12)
            hour += 12;
        if (period == "AM" && hour == 12)
            hour = 0;

        return hour * 60 + minute;
    }
}
